crates/bevy_ecs/src/query/filter.rs

```diff
@@ -543,36 +562,36 @@ macro_rules! impl_flag_filter {
                         self.entity_table_rows = archetype.entity_table_rows().as_ptr();
                         // SAFE: archetype tables always exist
                         let table = tables.get_unchecked(archetype.table_id());
-                        self.table_flags = table
+                        self.table_counters = table
                             .get_column(state.component_id).unwrap()
-                            .get_flags_mut_ptr();
+                            .get_counters_mut_ptr();
                     }
                     StorageType::SparseSet => self.entities = archetype.entities().as_ptr(),
                 }
             }
 
             unsafe fn table_fetch(&mut self, table_row: usize) -> bool {
-                false $(|| (*self.table_flags.add(table_row)).contains($flags))+
+                false $(|| $counters(&*self.table_counters.add(table_row), self.system_counter, self.global_system_counter))+
```

Comment:
These unsafe logic lines deserve a thorough comment explaining how they work (as a group). They're important to understanding the code, and are in an unsafe block.